prometheus/prometheus.go

```diff
@@ -90,43 +140,104 @@ func (p *PrometheusSink) Collect(c chan<- prometheus.Metric) {
 	expire := p.expiration != 0
 	now := time.Now()
 	p.gauges.Range(func(k, v interface{}) bool {
-		if v != nil {
-			lastUpdate := v.(*PrometheusGauge).updatedAt
-			if expire && lastUpdate.Add(p.expiration).Before(now) {
+		if v == nil {
+			return true
+		}
+		g := v.(*gauge)
+		lastUpdate := g.updatedAt
+		if expire && lastUpdate.Add(p.expiration).Before(now) {
+			if g.canDelete {
 				p.gauges.Delete(k)
-			} else {
-				v.(*PrometheusGauge).Collect(c)
+				return true
 			}
+			// We have not observed the gauge this interval so we don't know its value.
+			g.Set(math.NaN())
 		}
+		g.Collect(c)
 		return true
 	})
 	p.summaries.Range(func(k, v interface{}) bool {
-		if v != nil {
-			lastUpdate := v.(*PrometheusSummary).updatedAt
-			if expire && lastUpdate.Add(p.expiration).Before(now) {
+		if v == nil {
+			return true
+		}
+		s := v.(*summary)
+		lastUpdate := s.updatedAt
+		if expire && lastUpdate.Add(p.expiration).Before(now) {
+			if s.canDelete {
 				p.summaries.Delete(k)
-			} else {
-				v.(*PrometheusSummary).Collect(c)
+				return true
 			}
+			// We have observed nothing in this interval.
+			s.Observe(math.NaN())
 		}
+		s.Collect(c)
 		return true
 	})
 	p.counters.Range(func(k, v interface{}) bool {
-		if v != nil {
-			lastUpdate := v.(*PrometheusCounter).updatedAt
-			if expire && lastUpdate.Add(p.expiration).Before(now) {
+		if v == nil {
+			return true
+		}
+		count := v.(*counter)
+		lastUpdate := count.updatedAt
+		if expire && lastUpdate.Add(p.expiration).Before(now) {
+			if count.canDelete {
 				p.counters.Delete(k)
-			} else {
-				v.(*PrometheusCounter).Collect(c)
+				return true
 			}
+			// Counters remain at their previous value when not observed, so we do not set it to NaN.
 		}
+		count.Collect(c)
 		return true
 	})
 }
 
+func initGauges(m *sync.Map, gauges []GaugeDefinition) {
+	for _, g := range gauges {
+		key, hash := flattenKey(g.Name, g.ConstLabels)
+		pG := prometheus.NewGauge(prometheus.GaugeOpts{
+			Name:        key,
+			Help:        g.Help,
+			ConstLabels: prometheusLabels(g.ConstLabels),
+		})
+		pG.Set(float64(0)) // Initialize at zero
```

Comment:
should gauges and summaries (which haven't been observed yet) be initialized to zero or `NaN` (as they do after expiration)?

Reply:
That is a good question - the initialization part here is a little redundant for counters and gauges because they start at zero in the go prometheus lib anyway. However, initializing summaries to 0 does create an artifact when the process starts and before they're emitted, where the line sits on the 0 like it's a counter or gauge. It's going to be better to delete all of these inits and just use the go prometheus client's defaults.